### DISCUSSION TITLE
Support for complex/composite keywords

### DISCUSSION THREAD

**Discussion by maiatobi** [photoprism/photoprism#3713] [Ideas]
### What does not work as expected?
Pictures that pre-processed by other software (ie digikam, ligthroom, etc) most likely will include keywords/tags/labels.
This keywords/tags/labels are imported into photoprism but it seems spaces are used as a separator so only individual words are keywords.

This is explained here:
[iptc:keywords can't have spaces in them. #2344](https://github.com/photoprism/photoprism/discussions/2344)

I think the scope of the explained requirement is too complex for what is actually needed:

All keywords are stored in the keyword table. What is the difference in storing a composite keyword that include one or more spaces?.

I made a test and inserted a keyword with value "KEY_A KEY_B" (without the quotations) in the keyword table, and a row in the keywords_photo table to associate it with a picture.

Photoprism is capable of finding the picture if I use the KEY_A but not if I use KEY_B. 

Photoprism is also capable of finding the picture if I use "KEY_A KEY_B"  (Including the quotations) in the search box. Which means that photoprism is not using double quote or single quote to search for exactly that term, is just ignoring the double quote and using OR on each word.

Even worse using KEY_A1 will also find the picture as it seems that numbers at the end of a word are ignored (So that also mens we cannot have keywords with alphanumeric characters?).

### What behavior do you expect?

I think it should be relatively simple to parse iptc:keywords or whatever field in IPTC/XMP are analysed to include spaces and use double quotes or single quotes in the search box to indicate we want to search for that exact text. A text with a space in the middle is not different that a text without space in the middle for a database instead of using "KEYWORD%" it can search "%KEYWORD%" so it will match if the keyword is contained not only if the keyword start with the search parameter like is done now.

thanks
PS One of the anoying things of the current system is that connecting words are created as keywords (ie "de", "la", "como" in spanish).

**Comment by graciousgrey** [comment]
Thanks for sharing your use case! 

PhotoPrism aggregates keywords from many different sources, many of which contain a lot of noise that would clutter the index, e.g. automatically generated filenames like da39a3ee5e6b4b0d3255bfef95601890afd80709.jpeg or metadata automatically generated by smartphones and added to images, for example internally used IDs. 

We plan to additionally offer a [tagging system](https://github.com/photoprism/photoprism/issues/1779) that will allow you to organize your pictures with arbitrary tags - that may also include spaces, numbers and emojis - so that you have complete freedom and your tags are not mixed with automatically extracted metadata.

**Comment by graciousgrey** [comment]
> PS One of the anoying things of the current system is that connecting words are created as keywords (ie "de", "la", "como" in spanish).

Do you have an example for this? From what source do you get those keywords?

**Reply by maiatobi** [reply]
Hi. 
Sure. In my picture I added a lightroom keyword "casa de mamá"  (spanish), also de caption/title is "cumpleaños de francisco" the keywords added by photoprism are:
brown, casa, cumpleaños, de, francisco, luis, mamá 

I sympathize with your intention of "PhotoPrism aggregates keywords from many different sources", but this is really a problem if the only thing you do is blindingly copying words. 

Even  if the common user will not use a separate photo manager (ie ligthroom, digiKam, etc) the fact that you are importing the ligthroom/digikam tags/keywords as simple words is a defect, because that is not the way they were created.  

The problem is compounded by the simple search engine used. 

I think it would be a good change if you choose to use a Full-Text Index in MariaDB.

You will simplify the database design as you will not need 1 row per word. You either will need 1 row per phrase or a single large text with all word you collected and let the database deal with the search.
You just need to also offer in the search box to recognize the MariaDB operators. 
I know you also support sqllite but I think if you really want to scale this to the size of number of pictures a regular user generates (not to mention groupd), probably sqllite is not a fit... plus is more work to maintain 2 databases :-). 

regards

**Reply by lastzero** [reply]
For the moment, the keywords are words in the common sense. I'm sorry that we can't offer you a better solution right now. A future version will support tags that can be any string, and will allow using Solr for the search index, which will solve your issues.

**Reply by maiatobi** [reply]
hi,
Please do not misunderstand me, I know engineering a piece of software is a delicate balance between supporting current customers, how much effort is to add changes (and how many people is available to work)  and how much complexity you add to the product and that is a decision only you guys can make. 
Solr sound good but again it will add a lot of complexity and probably effort. I am not sure how much effort and I do not know how much effort would be to change the system to use MariaDB FULL-TEXT indexes but looks to me you could get most of the requirements done with MariaDB for a fraction of the complexity and  effort of using Solr.

regards
